Fix md5 value and file size update in files metadata console

pli888 opened the pull request:
# Pull request for issue: #2137 

This is a pull request for the following functionalities:

* A fix to files metadata console, specifically the correct file size and md5 value of a file is updated even if other files with the same file name exist in other datasets.

## How to test?

### Dev environment

Spin up a local install of GigaDB:
```
$ ./up.sh
```

Two new dev datasets 102442 and 102722 have been added to allow manual testing of this PR. In both datasets, there is a file with the same file name called `BUSCO_full_table.tsv`:
```
$ docker-compose run --rm test psql -h database -p 5432 -U gigadb gigadb -c "select file.id, name, size, identifier from file inner join dataset on file.dataset_id = dataset.id where identifier = '102722' or identifier = '102442';"
   id   |         name         |   size    | identifier 
--------+----------------------+-----------+------------
 537944 | BUSCO_full_table.tsv |    373021 | 102442
 537955 | Ptyas_repeat.gff     | 242243729 | 102442
 551147 | BUSCO_full_table.tsv |    373021 | 102722
 551150 | Masurca_config.txt   |      8878 | 102722
(4 rows)
```

Also, confirm there are no md5 values for dataset 102722 and 102442 files:
```
$ docker-compose run --rm test psql -h database -p 5432 -U gigadb gigadb -c "select file.name, file_attributes.value from file, file_attributes where file.dataset_id = (select id from dataset where identifier = '102722') and file_attributes.attribute_id = 605 and file.id = file_attributes.file_id;"
 name | value 
------+-------
(0 rows)


$ docker-compose run --rm test psql -h database -p 5432 -U gigadb gigadb -c "select file.name, file_attributes.value from file, file_attributes where file.dataset_id = (select id from dataset where identifier = '102442') and file_attributes.attribute_id = 605 and file.id = file_attributes.file_id;"
 name | value 
------+-------
(0 rows)
```

Change directory into user dropbox:
```
$ cd gigadb/app/tools/files-metadata-console/tests/_data/dropbox/user201
```

Notice it contains 102722.filesizes and 102722.md5; both files contain 2 rows of metadata to update 2 files: `BUSCO_full_table.tsv` and `Masurca_config.txt` in dataset 102722. 

Run script to update file metadata using above two files:
```
$ ../../../../scripts/filesMetaToDb.sh 102722
Updating md5 checksum values as file attributes for 102722
Number of changes: 2
Updating file sizes for 102722
Number of changes: 2
Updated file metadata for 102722 in database
```

Query database to check if file sizes have been correctly updated for dataset 102722 in particular that for BUSCO_full_table.tsv:
```
$ cd ../../../../../../../..
$ docker-compose run --rm test psql -h database -p 5432 -U gigadb gigadb -c "select file.id, name, size, identifier from file inner join dataset on file.dataset_id = dataset.id where identifier = '102722' or identifier = '102442';"
   id   |         name         |   size    | identifier 
--------+----------------------+-----------+------------
 537944 | BUSCO_full_table.tsv |    373021 | 102442
 537955 | Ptyas_repeat.gff     | 242243729 | 102442
 551150 | Masurca_config.txt   |      8878 | 102722
 551147 | BUSCO_full_table.tsv |    999999 | 102722
(4 rows)
```

Before this PR, the BUSCO_full_table.tsv file from dataset 102442 would have been wrongly updated probably because it comes first in order of `id`.

Check md5 values have been updated in dataset 102722:
```
$ docker-compose run --rm test psql -h database -p 5432 -U gigadb gigadb -c "select file.name, file_attributes.value from file, file_attributes where file.dataset_id = (select id from dataset where identifier = '102722') and file_attributes.attribute_id = 605 and file.id = file_attributes.file_id;"
         name         |              value               
----------------------+----------------------------------
 BUSCO_full_table.tsv | 61d9dab96def1cf77f7e29bsiumai888
 Masurca_config.txt   | 2590e2216fc2de16ca972ece01bae2d0
(2 rows)
```

The md5 values for dataset 102442 files have not been updated:
```
$ docker-compose run --rm test psql -h database -p 5432 -U gigadb gigadb -c "select file.name, file_attributes.value from fi
le, file_attributes where file.dataset_id = (select id from dataset where identifier = '102442') and file_attributes.attribute_id = 605 and file.id = file_attributes.file_id;"
 name | value 
------+-------
(0 rows)
```

Before this PR, the BUSCO_full_table.tsv file from dataset 102442 would have had its md5 value wrongly updated.

### Running functional tests

Run update file size tests:
```
$ pwd
path/to/gigadb-website/gigadb/app/tools/files-metadata-console
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/functional/UpdateFileSizeCest.php
```

Run update md5 value tests:
```
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/functional/UpdateFileAttributeMd5ValueCest.php
```

### Staging environment

Deploy your staging server:
```
cd /path/to/gigadb-website/ops/infrastructure/envs/staging
# Copy terraform files to staging environment
../../../scripts/tf_init.sh --project gigascience/forks/<username>-gigadb-website --env staging --region <region> --ssh-key ~/.ssh/<name>.pem

AWS_PROFILE=<name> terraform plan
AWS_PROFILE=<name> terraform apply
AWS_PROFILE=<name> terraform refresh

../../../scripts/ansible_init.sh --env staging

env TF_KEY_NAME=private_ip OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook  -i ../../inventories webapp_playbook.yml -e="gigadb_env=staging"

env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml --extra-vars="gigadb_env=staging"

# Create curator user account on bastion server
ansible-playbook -i ../../inventories users_playbook.yml -e "newuser=peterl" -e "credentials_csv_path=~/Desktop/credentials-peterl.csv" -e "gigadb_env=staging"
```

Run GitLab CI/CD pipeline to deploy this PR branch onto your staging environment.

Your staging server will be running with production data from latest GigaDB database backup. Let's query information about datasets 102722 and and 102442 to provide a reference before using post upload tool that will update file size and md5 values for their files.

Log into bastion server for your staging environment as your unprivileged curator user:
```
$ ssh -i ~/.ssh/<filename>.pem <username>@<staging-bastion-ip-address>
```

Check file size of BUSCO_full_table.tsv from dataset 102772:

> [!TIP]
> You will be asked to enter your staging database password
```
[peterl@ip-10-99-0-18 ~]$ psql -h rds-server-staging-pli.c4kksj3h88ei.ap-south-1.rds.amazonaws.com -U gigadb -c "select id, name, size from file where dataset_id = (select id from dataset where identifier='102722') and name = 'BUSCO_full_table.tsv';"
Password for user gigadb: 
   id   |         name         |  size  
--------+----------------------+--------
 551147 | BUSCO_full_table.tsv | 373021
(1 row)
```

Also, check md5 value for BUSCO_full_table.tsv:
```
[peterl@ip-10-99-0-18 ~]$ psql -h rds-server-staging-pli.c4kksj3h88ei.ap-south-1.rds.amazonaws.com -U gigadb -c "select file.id, file.name, file_attributes.value from file, file_attributes where file.dataset_id = (select id from dataset where identifier='102722') and file_attributes.attribute_id=605 and file.id = file_attributes.file_id and file.name = 'BUSCO_full_table.tsv';"
Password for user gigadb: 
   id   |          name        |              value               
--------+----------------------+----------------------------------
 551147 | BUSCO_full_table.tsv | 61d9dab96def1cf77f7e29b9dadac5f3
(1 row)
```

Set up user dropbox and files for performing post upload operation:
```
# Change directory
[peterl@ip-10-99-0-18 ~]$ cd /share/dropbox

# Create folder and change directory
[peterl@ip-10-99-0-18 ~]$ mkdir user201
[peterl@ip-10-99-0-18 ~]$ cd user201

# Create metadata files
[peterl@ip-10-99-0-18 ~]$ echo -e '999999\t./BUSCO_full_table.tsv\n8878\t./Masurca_config.txt' > 102722.filesizes
[peterl@ip-10-99-0-18 ~]$ echo -e '61d9dab96def1cf77f7e29bsiumai888  ./BUSCO_full_table.tsv\n2590e2216fc2de16ca972ece01bae2d0  ./Masurca_config.txt' > 102722.md5
```

Run file metadata update:
```
[peterl@ip-10-99-0-18 ~]$ sudo filesMetaToDb 102722
Updating md5 checksum values as file attributes for 102722
Number of changes: 2
Updating file sizes for 102722
Number of changes: 2
Updated file metadata for 102722 in database
```

You will now see that BUSCO_full_table.tsv in dataset 102722 has had its file size updated to that found in 102722.filesizes. Previously, another file also called BUSCO_full_table.tsv from another dataset would have been updated.
```
[peterl@ip-10-99-0-18 ~]$ psql -h rds-server-staging-pli.c4kksj3h88ei.ap-south-1.rds.amazonaws.com -U gigadb -c "select id, name, size from file where dataset_id = (select id from dataset where identifier='102722') and name = 'BUSCO_full_table.tsv';"
Password for user gigadb: 
   id   |         name         |  size  
--------+----------------------+--------
 551147 | BUSCO_full_table.tsv | 999999
(1 row)
```

The md5 value for dataset 102722's BUSCO_full_table.tsv file has also been updated - it now includes a name of a famous dim sum:
```
[peterl@ip-10-99-0-18 ~]$ psql -h rds-server-staging-pli.c4kksj3h88ei.ap-south-1.rds.amazonaws.com -U gigadb -c "select file.id, file.name, file_attributes.value from file, file_attributes where file.dataset_id = (select id from dataset where identifier='102722') and file_attributes.attribute_id=605 and file.id = file_attributes.file_id and file.name = 'BUSCO_full_table.tsv';"
Password for user gigadb: 
   id   |         name         |              value               
--------+----------------------+----------------------------------
 551147 | BUSCO_full_table.tsv | 61d9dab96def1cf77f7e29bsiumai888
(1 row)

```

## How have functionalities been implemented?

Multiple conditions for querying database tables need to be appended together using [`andwhere`](https://www.yiiframework.com/doc/guide/1.1/en/database.query-builder) in `DatasetFilesUpdater` class. Previously, `where` was used which was the cause of the bug.

The other files that have been changed in this PR are mostly csv files containing dev and gigadb_testdata.

## Any changes to automated tests?

New functional tests have been created in `UpdateFileSizeCest` and `UpdateFileAttributeMd5ValueCest`.

Changes to csv files for gigadb_testdata were required for these tests.

A couple of functional and acceptance tests had to be updated because of new datasets in gigadb_testdata.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Added new metadata files for file sizes and MD5 checksums for dataset 102722.
  - Updated ignore settings to allow tracking of new metadata files.
- **Bug Fixes**
  - Improved file update logic to ensure changes are only applied to files within the correct dataset, even if filenames are duplicated across datasets.
- **Tests**
  - Introduced a new test to verify that file size updates are restricted to the correct dataset when files share the same name.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pli888 committed:
Add dataset 102442 into dev data

pli888 committed:
Add dataset 102722 into dev data

pli888 committed:
Use andwhere to append multiple conditions for database query

pli888 committed:
Update gitignore for files-metadata-console

pli888 committed:
Create new data for testing file-metadata-console

pli888 committed:
Add missing data and fix foreign key contraints in dev data

pli888 committed:
Add new data into gigadb_testdata

pli888 committed:
Add new data for functional test

pli888 committed:
Add new functional test

Check only BUSCO_full_table.tsv file in dataset 102722 has file
size updated, but not file with same name in dataset 102442.

pli888 committed:
Add functional test to check md5 value update

pli888 committed:
Add entry into CHANGELOG.md

pli888 commented:
@coderabbitai review

pli888 committed:
Fix api test

Two new datasets have been added so expected count has changed from
6 to 8.

pli888 committed:
Fix acceptance tests

The id of the latest dev dataset is 2966. Therefore, id of a new
dataset will be 2967.

kencho51 reviewed (commented):
Hi @pli888,

Here are the steps I used to setup my local env:
```
% cd gigadb-website
% ./up.sh
% cd gigadb/app/tools/files-metadata-console
% docker-compose run --rm configure
% docker-compose run --rm composer install
```

**praise:** The files-metadata-console tool is able to update the file size and the md5 values of a specified DOI, even the file name is the same as another DOI.

**praise:** The unit tests and the functional tests are passing.

**issue:** When I `bats tests/bats`, there are 5 failures:
```
 ✗ DOI provided
   (in test file tests/bats/md5.bats, line 32)
     `[ "${lines[0]}" = "Created 102480.md5" ]' failed
   102480.filesizes
   102480.md5
   Executing teardown code
   Deleting ./102480.md5
   Deleting ./102480.filesizes
   Deleting tests/_data/102480/102480.md5
   Deleting tests/_data/102480/102480.filesizes
 ✗ Confirm md5 and file size values
   (in test file tests/bats/md5.bats, line 42)
     `[ "$output" = '67d9336ca3b61384185dc665026a2325  ./analysis_data/Tree_file.txt' ]' failed
   102480.filesizes
   102480.md5
   Executing teardown code
   Deleting ./102480.md5
   Deleting ./102480.filesizes
   Deleting tests/_data/102480/102480.md5
   Deleting tests/_data/102480/102480.filesizes
 ✗ Execute md5.sh within container
   (in test file tests/bats/md5.bats, line 53)
     `docker-compose run --rm -w /gigadb/app/tools/files-metadata-console/tests/_data/102480 files-metadata-console ../../../scripts/md5.sh 102480' failed
 ✗ Test postUpload.sh
   Executing setup code
   Downloading Excel file for dataset 102498
   Executing teardown code
   (from function `setup' in test file tests/bats/postUpload.bats, line 19)
     `cd ../../../excel-spreadsheet-uploader || exit' failed
     % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                    Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  262k  100  262k    0     0  78588      0  0:00:03  0:00:03 --:--:--  122k
/Users/kencho/gigascience/gigadb-website/gigadb/app/tools/files-metadata-console/tests/bats/postUpload.bats: line 19: cd: ../../../excel-spreadsheet-uploader: No such file or directory
updateUrls.bats
 ✗ transform dataset and file URLs
```

**issue:** 
```
% ../../../../scripts/filesMetaToDb.sh 102722
Updating md5 checksum values as file attributes for 102722
WARN[0000] a network with name deployment_db-tier exists but was not created for project "files-metadata-console".
Set `external: true` to use an existing network 
```

**issue:** After uncommet the `external: true` in `gigadb/app/tools/files-metadata-console/docker-compose.yml`:

```
/gigadb/app/tools/files-metadata-console/tests/_data/dropbox/user201 (pr2345-peter-fix-md5-update$*%) % ../../../../scripts/filesMetaToDb.sh 102722
Updating md5 checksum values as file attributes for 102722
WARN[0000] /Users/kencho/gigascience/gigadb-website/gigadb/app/tools/files-metadata-console/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
[+] Building 3.1s (8/9)                                                                                                                                                                                       docker:desktop-linux
 => [files-metadata-console internal] load build definition from Dockerfile                                                                                                                                                   0.0s
 => => transferring dockerfile: 1.11kB                                                                                                                                                                                        0.0s
 => [files-metadata-console internal] load metadata for docker.io/library/php:8.2.3-cli-bullseye                                                                                                                              1.0s
 => [files-metadata-console internal] load .dockerignore                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                               0.0s
 => [files-metadata-console 1/6] FROM docker.io/library/php:8.2.3-cli-bullseye@sha256:2f41da28d8f683d4db7cecbbe0cfe2a2ec7ee93ec3de4a7db91d211423fcbaf5                                                                        0.0s
 => => resolve docker.io/library/php:8.2.3-cli-bullseye@sha256:2f41da28d8f683d4db7cecbbe0cfe2a2ec7ee93ec3de4a7db91d211423fcbaf5                                                                                               0.0s
 => CACHED [files-metadata-console 2/6] RUN apt-get update -yqq &&     apt-get install -y apt-utils &&     apt-get clean                                                                                                      0.0s
 => CACHED [files-metadata-console 3/6] RUN apt-get update -yq &&     apt-get install -y --no-install-recommends         libzip-dev         libpq-dev &&         apt-get clean                                                0.0s
 => CACHED [files-metadata-console 4/6] RUN docker-php-ext-install pdo_pgsql &&     docker-php-ext-install pgsql &&     apt-get install -y postgresql-client-13                                                               0.0s
 => ERROR [files-metadata-console 5/6] RUN curl -L -o gum_0.14.5_amd64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.14.5/gum_0.14.5_amd64.deb' &&     dpkg --install gum_0.14.5_amd64.deb &&     rm gum_0.  2.0s
------                                                                                                                                                                                                                             
 > [files-metadata-console 5/6] RUN curl -L -o gum_0.14.5_amd64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.14.5/gum_0.14.5_amd64.deb' &&     dpkg --install gum_0.14.5_amd64.deb &&     rm gum_0.14.5_amd64.deb:                                                                                                                                                                                                                                  
0.112   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                                                              
0.112                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 4383k  100 4383k    0     0  2382k      0  0:00:01  0:00:01 --:--:-- 8585k
1.990 dpkg: error processing archive gum_0.14.5_amd64.deb (--install):
1.990  package architecture (amd64) does not match system (arm64)
1.996 Errors were encountered while processing:
1.996  gum_0.14.5_amd64.deb
------
failed to solve: process "/bin/sh -c curl -L -o gum_0.14.5_amd64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.14.5/gum_0.14.5_amd64.deb' &&     dpkg --install gum_0.14.5_amd64.deb &&     rm gum_0.14.5_amd64.deb" did not complete successfully: exit code: 1

```

**suggestion:** The cmd `% ../../../../scripts/filesMetaToDb.sh 102722` can be executed successfully after I have updated the dockerfile as:
```
# Install gum
RUN curl -L -o gum_0.14.5_arm64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.14.5/gum_0.14.5_arm64.deb' && \
    dpkg --install gum_0.14.5_arm64.deb && \
    rm gum_0.14.5_arm64.deb
```

pli888 committed:
Upgrade gum to latest version

pli888 committed:
Use files-metadata-console container for running bats tests

pli888 committed:
Fix change directory to upload tool folder

pli888 committed:
Fix 2 broken acceptance tests

These 2 search acceptance tests were broken because of new
datasets in dev data.

pli888 committed:
Fix bats test

updateUrls.bats was broken due to new datasets in dev data.

pli888 committed:
Reset dev database in teardown

pli888 commented:
@kencho51 

>issue: When I bats tests/bats, there are 5 failures:

Bats tests in files metadata console have been improved or fixed:

* Used files metadata console tool via container.
* Improved teardown code
* Fixed assertions due to new datasets in dev data
 
All bats test should run and pass now.

>% ../../../../scripts/filesMetaToDb.sh 102722
Updating md5 checksum values as file attributes for 102722
WARN[0000] a network with name deployment_db-tier exists but was not created for project "files-metadata-console".
Set `external: true` to use an existing network

This problem is already alluded to in docker-compose.yml for files metadata console.

> issue: After uncommet the external: true in gigadb/app/tools/files-metadata-console/docker-compose.yml:

I've updated the version of gum so this problem should not be seen now.

kencho51 commented:
@pli888 

**question:** Should we update the `gigadb-website/gigadb/app/tools/files-metadata-console/docker-compose.yml` ? Because  I still had to uncomment `external: true` otherwise I will get this error when I try to `% ../../../../scripts/filesMetaToDb.sh 102722` 
```
WARN[0000] a network with name deployment_db-tier exists but was not created for project "files-metadata-console".
Set external: true to use an existing network
```

**Issue:** When I tried to execute `filesMetaToDb.sh`, I still got this error:

```
% cd gigadb/app/tools/files-metadata-console/tests/_data/dropbox/user201
% ../../../../scripts/filesMetaToDb.sh 102722 
Updating md5 checksum values as file attributes for 102722
[+] Building 3.6s (9/10)                                                                                                                                                                                      docker:desktop-linux
 => [files-metadata-console internal] load build definition from Dockerfile                                                                                                                                                   0.0s
 => => transferring dockerfile: 1.11kB                                                                                                                                                                                        0.0s
 => [files-metadata-console internal] load metadata for docker.io/library/php:8.2.3-cli-bullseye                                                                                                                              1.6s
 => [files-metadata-console auth] library/php:pull token for registry-1.docker.io                                                                                                                                             0.0s
 => [files-metadata-console internal] load .dockerignore                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                               0.0s
 => [files-metadata-console 1/6] FROM docker.io/library/php:8.2.3-cli-bullseye@sha256:2f41da28d8f683d4db7cecbbe0cfe2a2ec7ee93ec3de4a7db91d211423fcbaf5                                                                        0.0s
 => => resolve docker.io/library/php:8.2.3-cli-bullseye@sha256:2f41da28d8f683d4db7cecbbe0cfe2a2ec7ee93ec3de4a7db91d211423fcbaf5                                                                                               0.0s
 => CACHED [files-metadata-console 2/6] RUN apt-get update -yqq &&     apt-get install -y apt-utils &&     apt-get clean                                                                                                      0.0s
 => CACHED [files-metadata-console 3/6] RUN apt-get update -yq &&     apt-get install -y --no-install-recommends         libzip-dev         libpq-dev &&         apt-get clean                                                0.0s
 => CACHED [files-metadata-console 4/6] RUN docker-php-ext-install pdo_pgsql &&     docker-php-ext-install pgsql &&     apt-get install -y postgresql-client-13                                                               0.0s
 => ERROR [files-metadata-console 5/6] RUN curl -L -o gum_0.16.1_amd64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.16.1/gum_0.16.1_amd64.deb' &&     dpkg --install gum_0.16.1_amd64.deb &&     rm gum_0.  1.9s
------                                                                                                                                                                                                                             
 > [files-metadata-console 5/6] RUN curl -L -o gum_0.16.1_amd64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.16.1/gum_0.16.1_amd64.deb' &&     dpkg --install gum_0.16.1_amd64.deb &&     rm gum_0.16.1_amd64.deb:                                                                                                                                                                                                                                  
0.162   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                                                              
0.162                                  Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 4606k  100 4606k    0     0  2801k      0  0:00:01  0:00:01 --:--:-- 13.7M
1.843 dpkg: error processing archive gum_0.16.1_amd64.deb (--install):
1.843  package architecture (amd64) does not match system (arm64)
1.849 Errors were encountered while processing:
1.849  gum_0.16.1_amd64.deb
------
failed to solve: process "/bin/sh -c curl -L -o gum_0.16.1_amd64.deb 'https://github.com/charmbracelet/gum/releases/download/v0.16.1/gum_0.16.1_amd64.deb' &&     dpkg --install gum_0.16.1_amd64.deb &&     rm gum_0.16.1_amd64.deb" did not complete successfully: exit code: 1
```

**suggestion:** I believe the reason for this is the gum version is hardcoded to `amd64` architecture, but my mac is with Apple chip that needs `arm64` version. Maybe update the dockerfile to install gum using apt as below:
```
sudo mkdir -p /etc/apt/keyrings
curl -fsSL https://repo.charm.sh/apt/gpg.key | sudo gpg --dearmor -o /etc/apt/keyrings/charm.gpg
echo "deb [signed-by=/etc/apt/keyrings/charm.gpg] https://repo.charm.sh/apt/ * *" | sudo tee /etc/apt/sources.list.d/charm.list
sudo apt update && sudo apt install gum
```

> Bats tests in files metadata console have been improved or fixed:

**issue:** The bats tests are still not passing, this time has 2 failuers:
```
% cd gigadb/app/tools/files-metadata-console
% bats tests/bats
md5.bats
 ✗ Execute md5.sh within container
   (in test file tests/bats/md5.bats, line 18)
     `docker-compose run --rm -w /gigadb/app/tools/files-metadata-console/tests/_data/dropbox/user4 files-metadata-console ../../../../scripts/md5.sh 102480' failed
   time="2025-06-20T15:36:32+08:00" level=warning msg="/Users/kencho/gigascience/gigadb-website/gigadb/app/tools/files-metadata-console/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion"
   Executing teardown code
   Deleting tests/_data/dropbox/user4/102480.md5
   Deleting tests/_data/dropbox/user4/102480.filesizes
postUpload.bats
 ✗ Test postUpload.sh
   Executing setup code
   Downloading Excel file for dataset 102498
   Ingesting Excel file for dataset 102498
   Executing test postUpload.sh --doi 102498 --dropbox user5
   Executing teardown code
Resetting dev database...
   (in test file tests/bats/postUpload.bats, line 65)
     `[ -f "${TOOLS_DIR}/files-metadata-console/tests/_data/dropbox/user5/102498.filesizes" ]' failed
     % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                    Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  262k  100  262k    0     0   152k      0  0:00:01  0:00:01 --:--:--  616k
...
12 tests, 2 failures
```

pli888 committed:
Add is_publishable column and data

The is_publishable column and data were missed out from rebase so
adding them in here in this commit.

pli888 committed:
Add check dataset type

This was missed off when rebasing with develop branch.

pli888 committed:
Update step to install gum

To resolve problem with Apple chip Macs that need arm64 version.

pli888 committed:
Fix foreign key association for curator_id

pli888 committed:
Update curator_id for dataset 102722 in dev data

pli888 committed:
Fix external: true warning message

Fix is provided by creating a docker-compose.override.yml file.

pli888 committed:
Remove version attribute

This will stop obselete warning message for this attribute.

pli888 committed:
Update readme documentation

pli888 committed:
Add back in version attribute

All docker-compose yml files need to have version attribute removed
otherwise Gitlab says there is version mismatch in
docker-compose.ci.yml.

pli888 commented:
@kencho51 

>question: Should we update the gigadb-website/gigadb/app/tools/files-metadata-console/docker-compose.yml ? Because I still had to uncomment external: true otherwise I will get this error when I try to % ../../../../scripts/filesMetaToDb.sh 102722

The `configure` script now includes a step to create a `docker-compose.override.yml` file which will resolve this problem. This problem is also documented in README.md.

>Issue: When I tried to execute filesMetaToDb.sh, I still got this error:

`Dockerfile` has been updated with a new way of installing gum which will avoid this problem. I have tested this is working on my laptop which can replicate this problem.

>issue: The bats tests are still not passing, this time has 2 failuers:

This problem has been fixed:
```
> bats tests/bats
 ✓ Show usage when no arguments passed
 ✓ Non existing file list
 ✓ Non existing search directory
 ✓ Empty file list
 ✓ non empty and existing file list and search directory passed to the command
 ✓ when the file list has blank lines in it
 ✓ Dot files are picked up in both directions
 ✓ Default non verbose mode only shows discrepancies
 ✓ Verbose mode shows all files and whether a match was found or not
 ✓ Execute md5.sh within container
 ✓ Test postUpload.sh
   Executing setup code
   Downloading Excel file for dataset 102498
   Ingesting Excel file for dataset 102498
   Executing test postUpload.sh --doi 102498 --dropbox user5
   Executing teardown code
Resetting dev database...
 ✓ transform dataset and file URLs
Resetting dev database...

12 tests, 0 failures
```

kencho51 reviewed (changes requested):
Hi @pli888,

**praise:** All my issues have been addressed and it is working in my dev. 

**issue:** The acceptance test is failing in my dev and also in gitlab pipeline, as you can see at https://gitlab.com/gigascience/forks/kencho-gigadb-website/-/jobs/10471681104

**question:** Why don't we uncomment `external: true` in the docker-compose.yml file, but creating `docker-compose.override.yml` to achieve the same goal?

pli888 committed:
Fix broken acceptance tests

When creating a new dataset from dev data, the new dataset will
have id 2967.

pli888 commented:
>issue: The acceptance test is failing in my dev and also in gitlab pipeline, as you can see at https://gitlab.com/gigascience/forks/kencho-gigadb-website/-/jobs/10471681104

The 3 broken acceptance tests have been fixed now.

>question: Why don't we uncomment external: true in the docker-compose.yml file, but creating docker-compose.override.yml to achieve the same goal?

If you uncomment uncomment `external: true` in the docker-compose.yml file, the pipeline will break in the `FilesMetaDataConsoleTestBuild` step. IIRC, this is due to using an old version of docker compose in Gitlab pipeline Therefore, we use an override file to bypass this problem in dev environment.

kencho51 reviewed (approved):
Hi @pli888,

**praise:** Refactoring the existing query using `andWhere` makes it specific, so no random update would happen.

**praise:** The dockerfile is updated to installed `gum` without hardcoded chip architecture.

**praise:** All tests are passing in my dev and also my gitlab pipeline.

Since all my queries have been addressed, happy to approve this PR.

rija reviewed (changes requested):
Hi @pli888 

**issue**: `filesMetadaToDb.sh` assumes that our project checkout is always called `gigadb-website`, which is not always true. More specifically in my case,  I keep checkout of everyone's fork with a different name (in your case my checkout is called `peter-gigadb`).
So the script trigger the error in line 63
**suggestion**: remove the requirement on the local git checkout to be called `gigadb-website`

**question**: related to the above, I've got the following error
```
../../../../scripts/filesMetaToDb.sh 102722                                                                                             (fix-md5-update) 17:44
../../../../scripts/filesMetaToDb.sh: line 64: err: command not found
```
Seems like it cannot find the `err` command. 
Is it bash version problem, or a dependency that need to be installed?

rija reviewed (changes requested):
Hi @pli888, 

**praise**: all tests pass on the Gitlab pipeline

**issue**: `filesMetadaToDb.sh` assumes that our local project checkout is always called `gigadb-website`, which is not always true. More specifically in my case,  I keep checkout of everyone's fork with a different name (in your case my checkout is called `peter-gigadb`).
So the script trigger the error in line 63
**suggestion**: remove the requirement on the local git checkout to be called `gigadb-website`

**question**: related to the above, I've got the following error
```
../../../../scripts/filesMetaToDb.sh 102722                                                                                             (fix-md5-update) 17:44
../../../../scripts/filesMetaToDb.sh: line 64: err: command not found
```
Seems like it cannot find the `err` command. 
Is it bash version problem, or a dependency that need to be installed?

**praise**: local test `tests/bats/md5.bats` pass
**issue**: local test `tests/bats/postUpload.bats` fails during the "Resetting dev database" phase, but running `./ops/scripts/setup_devdb.sh dev` independently succeeds and return success

**praise**: on my AWS deployment, the functionality works as described
